File: app/Packages/ONTDesignSystem/Sources/ONTDesignSystem/Tokens/ONTMetrics.swift
import SwiftUI

/// L'échelle d'espacement.
///
/// Des `@ScaledMetric` relatifs à `.body`, donc **ils grandissent avec Dynamic
/// Type** — le réglage système de taille de police. Pour une liseuse, et pour
/// un lectorat qui n'a pas vingt ans, une grille en points fixes est un défaut
/// d'accessibilité, pas un détail de style.
///
/// N'employer ces jetons que là où on aurait écrit un nombre au jugé. Pour le
/// reste, `.padding()` sans argument : SwiftUI tient déjà compte de la classe
/// de taille et du conteneur.
public struct ONTSpacing: DynamicProperty {
    // **Deux mises à l'échelle, une par plateforme.**
    //
    // `@ScaledMetric` suit le curseur du système — et ne fait rien sur macOS,
    // qui n'a pas de Dynamic Type (mesuré, voir `ONTEchelleUI`). `ONTUI.points`
    // rend donc la valeur telle quelle sur iOS, où elle est déjà mise à
    // l'échelle, et la multiplie par le facteur sur le Mac, où rien ne l'a
    // touchée. Un texte qui grandit dans des marges figées se serre contre
    // elles ; les deux doivent bouger ensemble.
    @ScaledMetric(relativeTo: .body) private var xsBrut: CGFloat = 4
    @ScaledMetric(relativeTo: .body) private var sBrut: CGFloat = 8
    @ScaledMetric(relativeTo: .body) private var mBrut: CGFloat = 12
    @ScaledMetric(relativeTo: .body) private var lBrut: CGFloat = 16
    @ScaledMetric(relativeTo: .body) private var pageBrut: CGFloat = 22
    @ScaledMetric(relativeTo: .body) private var xlBrut: CGFloat = 24
    @ScaledMetric(relativeTo: .body) private var xxlBrut: CGFloat = 32

    /// 4 pt — l'écart le plus serré : icône et libellé, intérieur d'une pastille.
    @MainActor public var xs: CGFloat { ONTUI.points(xsBrut) }
    /// 8 pt — un groupe lié : un titre au-dessus de son sous-titre.
    @MainActor public var s: CGFloat { ONTUI.points(sBrut) }
    /// 12 pt — marges d'une ligne, intérieur d'une carte.
    @MainActor public var m: CGFloat { ONTUI.points(mBrut) }
    /// 16 pt — la marge par défaut d'une vue.
    @MainActor public var l: CGFloat { ONTUI.points(lBrut) }
    /// 22 pt — les marges latérales d'une page de lecture.
    @MainActor public var page: CGFloat { ONTUI.points(pageBrut) }
    /// 24 pt — séparation entre sections.
    @MainActor public var xl: CGFloat { ONTUI.points(xlBrut) }
    /// 32 pt — respiration en tête d'écran.
    @MainActor public var xxl: CGFloat { ONTUI.points(xxlBrut) }

    public init() {}
}

/// Une taille en points, mise à l'échelle du curseur système.
///
/// `ONTSpacing` couvre les écarts, dont les valeurs sont peu nombreuses et se
/// nomment. Restait ce que l'on ne peut pas mettre en jetons : le corps d'un
/// `.system(size:)`, le côté d'un symbole SF, le minimum d'une case de grille —
/// des nombres dictés par un dessin précis, différents à chaque endroit.
///
/// Écrits en dur, ils ne bougent pas d'un pouce quand le lecteur monte
/// « Taille du texte » dans les réglages. C'est le piège de SwiftUI :
/// `Font.custom(_:size:)` suit ce curseur d'office, `.system(size:)` **non**.
/// Une vue qui mélange les deux voit sa hiérarchie se défaire dès que le
/// curseur bouge — le texte grandit, les étiquettes et les icônes restent.
///
/// ```swift
/// private var echelle = ONTScaled()
/// ...
/// Image(systemName: "xmark")
///     .font(.system(size: echelle(12), weight: .bold))
///     .frame(width: echelle(28), height: echelle(28))
/// ```
///
/// Le corps **et** le cadre qui le contient, toujours ensemble : un symbole
/// qui grandit dans une pastille figée en déborde.
public struct ONTScaled: DynamicProperty {
    @ScaledMetric(relativeTo: .body) private var facteur: CGFloat = 1

    public init() {}

    /// Voir `ONTSpacing` : sur le Mac, `@ScaledMetric` ne bouge pas, et c'est
    /// le facteur d'interface qui commande.
    @MainActor public func callAsFunction(_ points: CGFloat) -> CGFloat {
        ONTUI.points(points * facteur)
    }
}

/// Les rayons de courbure.
public enum ONTRadius {
    /// Pastilles et étiquettes.
    public static let pill: CGFloat = 999
    /// Surlignage d'un verset — juste assez pour adoucir l'angle.
    public static let highlight: CGFloat = 6
    /// Blocs secondaires.
    public static let block: CGFloat = 18
    /// Cartes de premier plan.
    public static let card: CGFloat = 22
}

/// La largeur de lecture.
public enum ONTLayout {
    /// Au-delà, une ligne devient trop longue pour que l'œil retrouve le
    /// début de la suivante. Vaut surtout sur iPad.
    public static let readingWidth: CGFloat = 700

    /// La largeur d'une **page** — listes, cartes, réglages.
    ///
    /// Plus large que la mesure du texte suivi, et c'est délibéré : une liste
    /// ne se lit pas comme une phrase. L'œil n'y court pas d'un bout à l'autre,
    /// il saute d'un intitulé à sa valeur. La serrer à la mesure de la prose la
    /// tassait sans rien gagner en lisibilité.
    public static let pageWidth: CGFloat = 850

    /// La largeur d'une carte — celle qu'elle a sur iPhone.
    ///
    /// Un iPhone de 402 points moins ses deux marges de vingt : la carte du
    /// verset y fait 363 × 232, soit un rapport de 1,56. Étalée sur la colonne
    /// de l'iPad, elle passait à 803 × 235 — un rapport de 3,4, une bande. Le
    /// verset n'y tenait plus que sur deux lignes qui traversaient l'écran.
    ///
    /// La carte garde donc sa mesure d'un appareil à l'autre. C'est aussi ce
    /// qui la maintient jumelle de la pastille du widget, qui ne s'étire pas
    /// non plus.
    public static let cardWidth: CGFloat = 362
}

File: android/ontdesignsystem/src/main/kotlin/com/labibleont/ont/designsystem/metrics/ONTMetrics.kt
package com.labibleont.ont.designsystem.metrics

import androidx.compose.runtime.Composable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * L'échelle d'espacement.
 *
 * ## Elle grandit avec le curseur d'accessibilité
 *
 * Les valeurs sont multipliées par `fontScale`, le réglage système de taille de
 * texte. Pour une liseuse — et pour un lectorat qui n'a pas vingt ans — une
 * grille en points fixes est un **défaut d'accessibilité**, pas un détail de
 * style : le texte grandit, les marges restent, et la page se serre jusqu'à ce
 * que les lignes se touchent.
 *
 * C'est le pendant des `@ScaledMetric` d'iOS. Compose ne les a pas : `sp` suit
 * le curseur, `dp` non. Il faut donc l'appliquer soi-même, et c'est le genre de
 * chose qu'on n'ajoute jamais après coup.
 *
 * N'employer ces jetons que là où on aurait écrit un nombre au jugé.
 */
public class ONTSpacing internal constructor(private val facteur: Float) {
    /** 4 — l'écart le plus serré : icône et libellé, intérieur d'une pastille. */
    public val xs: Dp get() = 4.dp * facteur

    /** 8 — un groupe lié : un titre au-dessus de son sous-titre. */
    public val s: Dp get() = 8.dp * facteur

    /** 12 — marges d'une ligne, intérieur d'une carte. */
    public val m: Dp get() = 12.dp * facteur

    /** 16 — la marge par défaut d'une vue. */
    public val l: Dp get() = 16.dp * facteur

    /** 22 — les marges latérales d'une page de lecture. */
    public val page: Dp get() = 22.dp * facteur

    /** 24 — séparation entre sections. */
    public val xl: Dp get() = 24.dp * facteur

    /** 32 — respiration en tête d'écran. */
    public val xxl: Dp get() = 32.dp * facteur

    /**
     * Une mesure quelconque, mise à l'échelle.
     *
     * Pour ce qui ne se met pas en jetons : le côté d'une icône, le minimum
     * d'une case. Le corps **et** le cadre qui le contient, toujours ensemble —
     * un symbole qui grandit dans une pastille figée en déborde.
     */
    public operator fun invoke(points: Int): Dp = points.dp * facteur
}

/**
 * L'échelle d'espacement, mesurée sur le texte qu'elle accompagne.
 *
 * ## Pourquoi pas `fontScale`
 *
 * On multipliait les `dp` par `LocalDensity.current.fontScale`. C'était juste
 * jusqu'à Android 14, qui a rendu la mise à l'échelle du texte **non
 * linéaire** — et pas d'une seule façon : le facteur dépend de la taille.
 * Mesuré sur `ONT-Pixel9`, curseur système au maximum :
 *
 * ```
 * fontScale=2.00  |  11sp ×2.000  13sp ×1.923  16sp ×1.750  22sp ×1.591  34sp ×1.222
 * fontScale=1.50  |  11sp ×1.500  13sp ×1.538  16sp ×1.438  22sp ×1.227  34sp ×1.000
 * ```
 *
 * Un petit texte grossit pleinement, un grand titre presque pas — à 1,5, un
 * titre de 34 sp ne bouge pas d'un point. C'est délibéré : au bout du curseur,
 * ce qui doit devenir lisible est le corps, pas les titres, et l'échelle
 * typographique se comprime.
 *
 * Conséquence : `fontScale` annonce 2,0 quand le corps ne grossit que de 1,75.
 * Des marges multipliées par 2,0 autour d'un texte grossi de 1,75 se
 * désaccordent de 14 % — précisément au réglage employé par ceux qui montent
 * le curseur parce qu'ils ne voient pas autrement. La documentation d'Android
 * le dit sans détour : `fontScale` n'est plus qu'indicatif.
 *
 * ## Ce qu'on fait à la place
 *
 * On ne modélise pas la courbe — ni logarithme, ni table recopiée, qui
 * divergeraient à la première révision d'Android. **On demande.**
 * `taille.sp.toDp()` fait passer la conversion réelle, courbe comprise, et le
 * rapport obtenu est le facteur exact appliqué à ce texte-là.
 *
 * Sous Android 14 la conversion est linéaire, et le rapport rend `fontScale` :
 * aucun test de version à écrire.
 *
 * @param taille la taille du texte, en `sp`, que cet espacement entoure.
 */
@Composable
@ReadOnlyComposable
public fun ontSpacingPour(taille: Float): ONTSpacing {
    val densite = LocalDensity.current
    val rendu = with(densite) { taille.sp.toDp().value }
    return ONTSpacing(rendu / taille)
}

/**
 * L'échelle d'espacement du mobilier.
 *
 * Calée sur 16 sp — `bodyLarge`, le corps du mobilier. Une page de lecture
 * doit employer [ontSpacingPour] avec la taille choisie par le lecteur : ses
 * marges accompagnent son texte à lui, qui peut être à deux fois cette
 * taille-ci.
 */
public val ontSpacing: ONTSpacing
    @Composable @ReadOnlyComposable
    get() = ontSpacingPour(16f)

/** Les rayons de courbure. */
public object ONTRadius {
    /** Pastilles et étiquettes. */
    public val pill: Dp = 999.dp

    /** Surlignage d'un verset — juste assez pour adoucir l'angle. */
    public val highlight: Dp = 6.dp

    /** Blocs secondaires. */
    public val block: Dp = 18.dp

    /** Cartes de premier plan. */
    public val card: Dp = 22.dp
}

/** La largeur de lecture. */
public object ONTLayout {
    /**
     * Au-delà, une ligne devient trop longue pour que l'œil retrouve le début
     * de la suivante. Vaut surtout sur tablette, où rien ne limiterait sinon.
     */
    public val readingWidth: Dp = 700.dp

    /**
     * La largeur d'une **page** — listes, cartes, réglages.
     *
     * Plus large que la mesure du texte suivi, et c'est délibéré : une liste ne
     * se lit pas comme une phrase. L'œil n'y court pas d'un bout à l'autre, il
     * saute d'un intitulé à sa valeur.
     */
    public val pageWidth: Dp = 850.dp

    /**
     * La largeur d'une carte — celle qu'elle a sur téléphone.
     *
     * Étalée sur la colonne d'une tablette, la carte du verset passait d'un
     * rapport de 1,56 à 3,4 — une bande, où le verset ne tenait plus que sur
     * deux lignes traversant l'écran. Elle garde donc sa mesure d'un appareil à
     * l'autre, ce qui la maintient aussi jumelle de la pastille du widget.
     */
    public val cardWidth: Dp = 362.dp
}
